hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -2205,6 +2272,167 @@ private void listStatusFindKeyInTableCache(
+        OzoneFileStatus fileStatus = getFileStatus(args, clientAddress);
+        if (fileStatus.isFile()) {
+          return Collections.singletonList(fileStatus);
+        }
+
+        // keyName is a directory
+        if (fileStatus.getKeyInfo() != null) {
+          seekKeyInDB = fileStatus.getKeyInfo().getObjectID()
+                  + OZONE_URI_DELIMITER;
+          prefixKeyInDB = fileStatus.getKeyInfo().getObjectID();
+        } else {
+          String bucketKey = metadataManager.getBucketKey(volumeName,
+                  bucketName);
+          OmBucketInfo omBucketInfo =
+                  metadataManager.getBucketTable().get(bucketKey);
+          seekKeyInDB = omBucketInfo.getObjectID()
+                  + OZONE_URI_DELIMITER;
+          prefixKeyInDB = omBucketInfo.getObjectID();
+        }
+      } else {
+        // startKey will be used in iterator seek and sets the beginning point
+        // for key traversal.
+        // key name will be used as parent ID where the user has requested to
+        // list the keys from.
+        OzoneFileStatus fileStatusInfo = getOzoneFileStatusV1(volumeName,
+                bucketName, startKey, false, null);
+        if (fileStatusInfo != null) {
+          prefixKeyInDB = fileStatusInfo.getKeyInfo().getParentObjectID();
+          seekKeyInDB = prefixKeyInDB + OZONE_URI_DELIMITER
+                  + fileStatusInfo.getKeyInfo().getFileName();
+        }
+      }
+
+      // Not required to search in TableCache because all the deleted keys
+      // are marked directly in directory table or in key table by breaking
+      // the pointer to its sub-dirs. So, there is no issue of inconsistency.
+      int countEntries = 0;
+      // Seek the given key in key table.
+      countEntries = getFilesFromDirectory(cacheKeyMap, seekKeyInDB,
+              prefixPath, countEntries, numEntries, prefixKeyInDB);
+      // Seek the given key in dir table.
+      Table dirTable = metadataManager.getDirectoryTable();
+      TableIterator<String, ? extends Table.KeyValue<String, OmDirectoryInfo>>
+              iterator = dirTable.iterator();
+
+      iterator.seek(seekKeyInDB);
+
+      while (iterator.hasNext() && numEntries - countEntries > 0) {
+        String entryInDb = iterator.key();
+        OmDirectoryInfo dirInfo = iterator.value().getValue();
+        if (!isImmediateChild(dirInfo.getParentObjectID(), prefixKeyInDB)) {
+          break;
+        }
+
+        if (recursive) {
+          // for recursive list all the entries
+          prefixPath = OMFileRequest.getAbsolutePath(prefixPath,
+                  dirInfo.getName());
+          OmKeyInfo omKeyInfo = OMFileRequest.getOmKeyInfo(volumeName,
+                  bucketName, dirInfo, prefixPath);
+          cacheKeyMap.put(entryInDb,
+                  new OzoneFileStatus(omKeyInfo, 0, true));
+          ++countEntries;
+          // files from this directory
+          seekKeyInDB = dirInfo.getObjectID() + OZONE_URI_DELIMITER;
```

Comment:
prefixKeyInDB should also updated to new dir objectID dirInfo.getObjectID()

Reply:
Done